File: src/commands/export.ts
import fs from 'fs';
import path from 'path';
import chalk from 'chalk';
import { Config, AppConfig } from '../types/config';
import { WingetService } from '../services/winget';

export async function exportCommand(outputPath: string): Promise<void> {
  const winget = new WingetService();

  console.log(chalk.cyan('\n📦 Fetching installed applications...\n'));

  const allApps = winget.getInstalledApps();

  if (allApps.length === 0) {
    console.log(chalk.yellow('No applications found.'));
    return;
  }

  // Filter out system apps, drivers, and runtimes
  const apps = allApps.filter(app => !winget.isSystemApp(app));
  const excludedCount = allApps.length - apps.length;

  console.log(chalk.cyan(`Found ${allApps.length} installed applications`));
  console.log(chalk.gray(`  Excluded ${excludedCount} system apps/drivers/runtimes`));
  console.log(chalk.cyan(`  Processing ${apps.length} user applications`));
  console.log(chalk.cyan('\nChecking availability in winget (this may take a while)...\n'));

  const configApps: AppConfig[] = [];
  let checked = 0;
  let availableCount = 0;
  let unavailableCount = 0;

  for (const app of apps) {
    checked++;
    process.stdout.write(`\r  Checking ${checked}/${apps.length}: ${app.name.substring(0, 40).padEnd(40)}`);

    // Apps with 'winget' source are definitely available
    const isFromWinget = app.source?.toLowerCase() === 'winget';
    let availableInWinget = isFromWinget;

    // For non-winget sources, check if available
    if (!isFromWinget && app.id) {
      availableInWinget = await winget.isAvailableInWinget(app.id);
    }

    if (availableInWinget) {
      availableCount++;
    } else {
      unavailableCount++;
    }

    const configApp: AppConfig = {
      id: app.id,
      name: app.name,
      version: 'latest',
      availableInWinget
    };

    if (!availableInWinget) {
      configApp.note = 'Not available via winget - install manually';
    }

    configApps.push(configApp);
  }

  console.log('\n');

  // Sort: available first, then by name
  configApps.sort((a, b) => {
    if (a.availableInWinget !== b.availableInWinget) {
      return a.availableInWinget ? -1 : 1;
    }
    return a.name.localeCompare(b.name);
  });

  const config: Config = {
    apps: configApps
  };

  // Write to file
  const absolutePath = path.resolve(outputPath);
  const dir = path.dirname(absolutePath);

  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  fs.writeFileSync(absolutePath, JSON.stringify(config, null, 2), 'utf-8');

  console.log(chalk.green(`✓ Config exported to: ${absolutePath}`));
  console.log(chalk.gray(`\nSummary:`));
  console.log(chalk.green(`  Available in winget: ${availableCount}`));
  console.log(chalk.yellow(`  Not available: ${unavailableCount}`));
  console.log(chalk.gray(`  Excluded (system): ${excludedCount}`));
  console.log(chalk.gray(`  Total exported: ${apps.length}\n`));

  if (unavailableCount > 0) {
    console.log(chalk.yellow('Note: Applications marked as unavailable will be skipped during installation.'));
    console.log(chalk.yellow('You can manually install them or find alternatives in winget.\n'));
  }
}

File: src/commands/export.rs
use std::fs;
use std::path::Path;

use anyhow::Result;

use crate::services::winget::WingetService;
use crate::types::{AppConfig, Config};

pub fn run(output_path: &str) -> Result<()> {
    let winget = WingetService::new();

    eprintln!("Fetching installed applications...");
    let all_apps = winget.get_installed_apps();

    if all_apps.is_empty() {
        println!("No applications found.");
        return Ok(());
    }

    let apps: Vec<_> = all_apps.iter().filter(|a| !winget.is_system_app(a)).collect();
    let excluded = all_apps.len() - apps.len();

    println!("Found {} installed applications", all_apps.len());
    println!("  Excluded {} system apps/drivers/runtimes", excluded);
    println!("  Processing {} user applications", apps.len());
    println!("\nChecking availability in winget (this may take a while)...\n");

    let mut config_apps: Vec<AppConfig> = vec![];
    let mut available_count = 0usize;
    let mut unavailable_count = 0usize;

    for (i, app) in apps.iter().enumerate() {
        let display = truncate(&app.name, 40);
        eprint!("\r  Checking {}/{}: {:<40}", i + 1, apps.len(), display);

        let from_winget = app.source.to_lowercase() == "winget";
        let available = if from_winget {
            true
        } else {
            winget.is_available_in_winget(&app.id)
        };

        if available {
            available_count += 1;
        } else {
            unavailable_count += 1;
        }

        let mut cfg = AppConfig {
            id: app.id.clone(),
            name: app.name.clone(),
            version: "latest".into(),
            available_in_winget: available,
            note: None,
        };
        if !available {
            cfg.note = Some("Not available via winget - install manually".into());
        }
        config_apps.push(cfg);
    }

    eprintln!();

    // Sort: available first, then by name
    config_apps.sort_by(|a, b| {
        b.available_in_winget
            .cmp(&a.available_in_winget)
            .then_with(|| a.name.cmp(&b.name))
    });

    let config = Config { apps: config_apps };

    let abs_path = Path::new(output_path);
    if let Some(parent) = abs_path.parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent)?;
        }
    }

    let json = serde_json::to_string_pretty(&config)?;
    fs::write(abs_path, json)?;

    println!("\nConfig exported to: {}", abs_path.display());
    println!("\nSummary:");
    println!("  Available in winget: {}", available_count);
    println!("  Not available:       {}", unavailable_count);
    println!("  Excluded (system):   {}", excluded);
    println!("  Total exported:      {}", apps.len());

    if unavailable_count > 0 {
        println!("\nNote: Apps marked unavailable will be skipped during installation.");
    }

    Ok(())
}

fn truncate(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        s.to_string()
    } else if max == 0 {
        String::new()
    } else {
        let mut out = s.chars().take(max.saturating_sub(1)).collect::<String>();
        out.push('…');
        out
    }
}
